WIP: Refactor systemd and use /etc/default/zfs if available

FransUrbo opened the pull request:
Rearrangement of some source directories.
    
* Move the systemd files into the contrib directory. This isn't really part of the ZFS/ZoL code, so it's more appropriate to put it in the contrib directory.

Use the /etc/{default,sysconfig}/zfs file in systemd service files (1of3).
    
* Instead of looing for the hardcoded @sysconfdir@/zfs/zpool.cache file, use the $ZFS_CACHE if/when set.
* Instead of using hardcoded /dev/disk/by-id dir in import, use the ZPOOL_IMPORT_PATH if/when set.
* Use ZPOOL_IMPORT_OPTS if/when applicable.

Use the /etc/{default,sysconfig}/zfs file in systemd service files (2of3).
    
* Use VERBOSE_MOUNT, DO_OVERLAY_MOUNTS and/or MOUNT_EXTRA_OPTIONS to <code>zfs mount</code> if/when applicable.

Use the /etc/{default,sysconfig}/zfs file in systemd service files (3of3).
    
* Use the ZED_ARGS variable when starting zed if/when applicable.


This whole PR is untested, so I'm tagging this **WIP**.

FransUrbo commented:
The first commit isn't really necessary, but I've been increasingly ... "annoyed" that there's stuff that isn't really part of ZoL in the tree. I think it makes more sense to have all that in the <code>contrib</code> directory (which we created it for). So this PR includes a move of the <code>systemd</code> directory into <code>contrib/</code>.

FransUrbo commented:
@behlendorf Do we have an (approximate) ETA on this?

behlendorf commented on configure.ac:
```diff
@@ -63,8 +63,6 @@ AC_CONFIG_FILES([
 	etc/init.d/Makefile
 	etc/zfs/Makefile
-	etc/systemd/Makefile
-	etc/systemd/system/Makefile
```
I don't think this belong in contrib.  This is code we support and maintain as part of the project so I'd like to leave where it is.

behlendorf commented on contrib/systemd/system/zfs-import-cache.service.in:
```diff
@@ -4,10 +4,11 @@ DefaultDependencies=no
 After=systemd-udev-settle.service
 After=cryptsetup.target
-ConditionPathExists=@sysconfdir@/zfs/zpool.cache
+ConditionPathExists=${ZFS_CACHE}
```
Should be `${ZPOOL_CACHE}`

FransUrbo replied:
Dang! That have been there for a very long time, and no one have noticed! Kind'a indicates how few that actually use the cache file!

behlendorf commented on contrib/systemd/system/zfs-import-cache.service.in:
```diff
@@ -4,10 +4,11 @@ DefaultDependencies=no
 RemainAfterExit=yes
 ExecStartPre=/sbin/modprobe zfs
-ExecStart=@sbindir@/zpool import -c @sysconfdir@/zfs/zpool.cache -aN
+ExecStart=@sbindir@/zpool import ${ZFS_CACHE:+-c "${ZFS_CACHE}"} -aN ${ZPOOL_IMPORT_OPTS}
```
/s/ZFS_CACHE/ZPOOL_CACHE/

behlendorf commented on contrib/systemd/system/zfs-import-cache.service.in:
```diff
@@ -4,10 +4,11 @@ DefaultDependencies=no
 ExecStartPre=/sbin/modprobe zfs
-ExecStart=@sbindir@/zpool import -c @sysconfdir@/zfs/zpool.cache -aN
+ExecStart=@sbindir@/zpool import ${ZFS_CACHE:+-c "${ZFS_CACHE}"} -aN ${ZPOOL_IMPORT_OPTS}
+EnvironmentFile=-@initconfdir@/zfs
```
Neat, I didn't know this existed but it's a nice approach.  Just one question, why `=-` instead of `=`?

behlendorf replied:
> PS. The prefixing dash makes it not fail if the file doesn't exist which I think is correct. It's at least uniform with the SYSV and initrd scripts.

I see you already explained, thanks.

behlendorf commented on contrib/systemd/system/zfs-import-scan.service.in:
```diff
@@ -4,10 +4,11 @@ DefaultDependencies=no
 After=systemd-udev-settle.service
 After=cryptsetup.target
-ConditionPathExists=!@sysconfdir@/zfs/zpool.cache
+ConditionPathExists=!${ZFS_CACHE}
```
/s/ZFS_CACHE/ZPOOL_CACHE/

behlendorf commented on contrib/systemd/system/zfs-mount.service.in:
```diff
@@ -12,4 +12,5 @@ Before=local-fs.target
 Type=oneshot
 RemainAfterExit=yes
-ExecStart=@sbindir@/zfs mount -a
+ExecStart=@sbindir@/zfs mount -a ${VERBOSE_MOUNT:+-v} ${DO_OVERLAY_MOUNTS:+-O} ${MOUNT_EXTRA_OPTIONS}
```
My only concern with this is that `VERBOSE_MOUNT="no"` and `DO_OVERLAY_MOUNT="no"`.  That means both these options will be enabled when they should be disabled.  How about limiting this to just `MOUNT_EXTRA_OPTIONS` which will work correctly.

behlendorf commented:
@FransUrbo sorry about completely neglecting this.  The gist of these changes look entirely reasonable to me.  If you can address the issues I mentioned and rebase this against master without the "Rearrangement of some source directories" patch I don't see why we can't get it tested and added as part of the other proposed systemd changes.

FransUrbo commented:
> @FransUrbo sorry about completely neglecting this.
> 
That's ok I guess, I've grown used to it :)
> If you can address the issues I mentioned and rebase this against master without the "Rearrangement of some source directories" patch I don't see why we can't get it tested and added as part of the other proposed systemd changes.
> 
In my opinion, everything that isn't part of the core ZFS code, should
go into contrib. It's technically the work of the downstream maintainers
to deal with this, not us.

But since there haven't been any, it made sense for us to do it. And 
we can keep doing it for everyone that prefer to install from source.

But I think contrib is correct in this case. However, we can leave that
discussion to later if you're adamant about this.

behlendorf commented:
Let's leave the contrib argument for latter and move forward on the rest of these changes.  :)

FransUrbo commented:
@behlendorf Ok, I've updated the PR with your remarks fixed. I updated the comments to, maybe you should merge them into one commit when/if accepting it.

FransUrbo commented:
Looking at this and your #4699 again, I think it's possible to merge the two import files into one and just update the documentation for /etc/{sysconfig,default}/zfs to indicate that one could use the <code>ZPOOL_IMPORT_OPTS</code> with **systemd** to choose between importing with cache or scan.

FransUrbo commented on etc/systemd/system/zfs-import.service.in:
```diff
@@ -1,16 +1,15 @@
 RemainAfterExit=yes
 EnvironmentFile=-@initconfdir@/zfs
 ExecStartPre=/sbin/modprobe zfs
-ExecStart=@sbindir@/zpool import ${ZPOOL_CACHE:+-c "${ZPOOL_CACHE}"} -aN ${ZPOOL_IMPORT_OPTS}
```
I'm pretty sure we can write the one line equivalence of 

Pseudo code:

if cache exists; then
    use cache file
else if ZPOOL_IMPORT_PATH is set; then
    use ZPOOL_IMPORT_PATH
fi

behlendorf replied:
Let's leave these two files separate for now.  That allows for each to enabled/disabled independently.

behlendorf commented on etc/systemd/system/zfs-import-cache.service.in:
```diff
@@ -5,10 +5,11 @@ Requires=systemd-udev-settle.service
 After=cryptsetup.target
 Before=dracut-mount.service
-ConditionPathExists=@sysconfdir@/zfs/zpool.cache
+ConditionPathExists=${ZPOOL_CACHE}
```
We should make this `${ZPOOL_CACHE:-@sysconfdir@/zfs/zpool.cache}` to support the case where there is no EnvironmentFile.

behlendorf commented on etc/systemd/system/zfs-import-cache.service.in:
```diff
@@ -5,10 +5,11 @@ Requires=systemd-udev-settle.service
+EnvironmentFile=-@initconfdir@/zfs
 ExecStartPre=/sbin/modprobe zfs
-ExecStart=@sbindir@/zpool import -c @sysconfdir@/zfs/zpool.cache -aN
+ExecStart=@sbindir@/zpool import ${ZPOOL_CACHE:+-c "${ZPOOL_CACHE}"} -aN ${ZPOOL_IMPORT_OPTS}
```
We'll need to do something similar here or we'll break all the systems which don't have this file installed.

FransUrbo replied:
```
EnvironmentFile=-@initconfdir@/zfs
Environment="ZPOOL_CACHE=${ZPOOL_CACHE:-@sysconfdir@/zfs/zpool.cache}"
```

Should work.

behlendorf commented:
When testing this under CentOS 7 I ran in to numerous problems.

* The environment variable substitutions aren't working, this may be because systemd does its own parsing of the unit files so normal shell syntax isn't supported.

* initconfdir replacement isn't defined and needs to be added to the Makefile.am: `-e 's,@initconfdir\@,$(DEFAULT_INITCONF_DIR),g'`.

* Variables appear not to be allowed in `ConditionPathExists`.

Adding this kind of support I think is a good idea but it needs more testing.

behlendorf commented:
Closing.  While I think this is a good idea the proposed patches just don't work.  I ran in to numerous problems as described in the [comment](https://github.com/zfsonlinux/zfs/pull/3884#issuecomment-222220980) above.  If someone would like to add this functionality they should open a new PR.
